share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,809 @@
+    | splitAt (strlen fullstring / 2) 
+  cat (laststr ++ firststr)                         
+```
+
+* Function dubmid takes a string, duplicates the middle character and gives out the string (Ex: hapy -> happy))
+```
+def dubmid fullstring =
+  def Pair firststr laststr = 
+    explode fullstring
+    | splitAt (strlen fullstring / 2)
+  def dupli = laststr | take 1
+  cat (firststr ++ dupli ++ laststr)
+```
+
+* Takes a string and then removes the last character in a string
+```
+wake -x '"been"| explode | take 3| cat'
+```
+
+* Replaces the first charcter with the character "c"
+```
+wake -x '"bad"| explode | drop 1| prepend "c" | cat' 
+```
+
+* Try out these examples which act on strings
+```
+wake -x '"south" | explode | take 1| cat'
+wake -x '"north"| explode | tail | take 1 | cat'
+wake -x '"east" | explode| reverse | take 1 |cat'
+wake -x '"west" | explode | reverse | tail |take 1| cat'
+```
+
+* Below are the set of examples which shows how to extract specific characters from a string
+```
+def first s = s | explode | take 1| cat
+def second s = s | explode | tail | take 1 | cat
+def third s = s | explode| tail | tail | take 1 |cat
+def fourth s =  s | explode | tail | tail | tail | take 1| cat
+def last s = s| explode | reverse |take 1 | cat
+```
+
+* Try out these different set of examples and capture the result
+```
+def roll s = "{fourth s}{first s}{second s}{third s}"
+def exch s = "{second s}{first s}{third s}{fourth s}"
+```
+
+* Try out these examples and record the results
+```
+def what fullstring = fullstring| roll | roll | roll| exch |  roll
+def fb fullstring = fullstring | roll |roll | roll
+def fc fullstring = fullstring | roll | exch |fb
+def fd fullstring = fullstring | exch | fc | exch
+```
+
+* This examples convert the string "seat" -> "eats"
+```
+wake -x 'cat((tail (explode "seat")) ++(take 1(explode "seat")))'
+```
+
+### Tuple
+A tuple is collection of items of different types. `get`,`set` and `edit` are the three inbuilt methods that can be used with the tuples.
+* Below example shows the declaration of a tuple. 
+```
+tuple Collection =
+  global IntValue : Integer
+  global DoubleValue : Double
+  global StringValue : String
+```
+
+`set<tuple name><field Name>` 
+`set` operation sets(writes) value to the tuple field
+`get<tuple name><field Name>` 
+`get` operation gets(reads) value to the tuple field
+`edit<tuple name><field Name>` 
+`edit` operation moidfies value to the tuple field
+
+### Map
+Map is a higher order function which takes a function and applies the same to every element in the list.
+
+* Below example shows how to declare a list of strings and apply map function to it. Map is a predefined function that maps to the every element in the list
+```
+def words = "ache", "vile", "amid", "evil", "ogre" , Nil
+def mapping words = words | map roll
+def mapping2 words = map exch words
+```
+
+### Higher Order function
+A function which does one of the following is a higher order function
```

Comment:
shouldn't this definition of a Higher Order function come before your use of it in defining `map` just above?

Reply:
Done